### PR TITLE
Add carbon metrics to get dashboard support

### PR DESCRIPTION
**Pull request by sahandilshan**
## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/CDCStatus.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package io.siddhi.extension.io.cdc.source.metrics;
```
Comment (deshankoswatte) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/CDCStatus.java:1]: Add the license header.

Comment (sahandilshan) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/CDCStatus.java:19]: Hi @deshankoswatte, This issue was fixed in e46325f81ffb02df800c3d47edd17907babdc83f commit

File: component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/MetricsUtils.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+ * under the License.
+ */
+package io.siddhi.extension.io.cdc.source.metrics;
+
+/**
+ * This class contains Util methods for the Metrics.
+ */
+public class MetricsUtils {
+
+    private static String getShortenURL(String jdbcURL) {
```
Comment (senthuran16) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/MetricsUtils.java:25]: Shall we say `getShortenedURL`

File: component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/MetricsUtils.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+            i++;
+            c = jdbcURL.charAt(i);
+        }
+        if (i == (n - 1)) {
+            return jdbcURL;
+        }
+        return jdbcURL.substring(0, i) + "..";
+    }
+
+    public static String getShortenJDBCURL(String url) {
```
Comment (senthuran16) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/MetricsUtils.java:45]: Shall we say `getShortenedJDBCURL`

File: component/src/main/java/io/siddhi/extension/io/cdc/source/CDCSource.java [hunk 4]
```diff
@@ -393,8 +410,20 @@ protected ServiceDeploymentInfo exposeServiceDeploymentInfo() {
...
         String tableName = optionHolder.validateAndGetOption(CDCSourceConstants.TABLE_NAME).getValue();
-        String siddhiAppName = siddhiAppContext.getName();
-
+        siddhiAppName = siddhiAppContext.getName();
+        this.siddhiAppContextExecutorService = siddhiAppContext.getExecutorService();
+        boolean isPrometheusReporterRunning = false;
+        if (MetricsDataHolder.getInstance().getMetricService() != null &&
+                MetricsDataHolder.getInstance().getMetricManagementService().isEnabled()) {
+            try {
+                if (MetricsDataHolder.getInstance().getMetricManagementService().isReporterRunning(
```
Comment (senthuran16) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/CDCSource.java:419]: Shall we replace this with 
```
isPrometheusReporterRunning = MetricsDataHolder.getInstance().getMetricManagementService().isReporterRunning(
                        "prometheus")
```

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java [hunk 5]
```diff
@@ -139,9 +155,14 @@ public void printEvent(Connection connection) {
                 handleEvent(detailsMap);
             }
         } catch (SQLException ex) {
+            if (metrics != null) {
+                isError = true;
```
Comment (CharukaK) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java:159]: Shouldn't we set the isError variable outside the metrics null check?

File: component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/PollingMetrics.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+            CDC_LAST_RECEIVED_TIME_MAP.put(cdcDatabase, lastPublishedTime);
+            lastReceivedTimeMetric();
+            setEventsInLastPollingMetric();
+            idleTimeMetric();
+        }
+    }
+
+    @Override
+    protected String getDatabaseURL() {
+            return host + "/" + tableName;
```
Comment (CharukaK) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/PollingMetrics.java:179]: use proper indents

File: component/pom.xml [hunk 7]
```diff
@@ -141,6 +141,11 @@
             <groupId>org.apache.kafka</groupId>
             <artifactId>connect-api</artifactId>
         </dependency>
+
```
Comment (CharukaK) [inline component/pom.xml:144]: remove unnecessary new line

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/WaitOnMissingRecordPollingStrategy.java [hunk 8]
```diff
@@ -145,13 +151,26 @@ public void poll() {
                         handleEvent(detailsMap);
                     }
                 } catch (SQLException e) {
+                    if (metrics != null) {
+                        isError = true;
```
Comment (CharukaK) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/WaitOnMissingRecordPollingStrategy.java:155]: set the isError variable outside the metrics null check

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java (hunk 5 above)
Comment (sahandilshan) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java:159]: `isError` is only needed when metrics service is enabled, That's why I'm setting it inside the if the condition

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/WaitOnMissingRecordPollingStrategy.java (hunk 8 above)
Comment (sahandilshan) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/WaitOnMissingRecordPollingStrategy.java:155]: Same as the above reason

File: component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java (hunk 5 above)
Comment (CharukaK) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/polling/strategies/DefaultPollingStrategy.java:159]: ack

File: component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/MetricsUtils.java (hunk 2 above)
Comment (sahandilshan) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/MetricsUtils.java]: Fix in 0964691b8ea7763a5dce1acce2fed88749011bcd

File: component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/MetricsUtils.java (hunk 3 above)
Comment (sahandilshan) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/MetricsUtils.java]: Fixed in 0964691b8ea7763a5dce1acce2fed88749011bcd

File: component/src/main/java/io/siddhi/extension/io/cdc/source/CDCSource.java (hunk 4 above)
Comment (sahandilshan) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/CDCSource.java]: Fixed in 9b553da8e0ca2f9d10e0868734562af8cfac30e7

File: pom.xml [hunk 9]
```diff
@@ -192,6 +192,13 @@
                 <artifactId>connect-api</artifactId>
                 <version>${kafka.connect.version}</version>
             </dependency>
+
+            <dependency>
+                <groupId>org.wso2.carbon.analytics</groupId>
+                <artifactId>org.wso2.carbon.si.metrics.core</artifactId>
+                <version>${carbon.analytics.version}</version>
+            </dependency>
+
```
[the remark this answers is not in the document]
Comment (sahandilshan) [reply pom.xml:207]: Fixed in 9b553da8e0ca2f9d10e0868734562af8cfac30e7

File: pom.xml [hunk 10]
```diff
@@ -264,6 +271,9 @@
         <kafka.connect.version>2.3.0</kafka.connect.version>
         <mongodb.connector.version>3.4.0</mongodb.connector.version>
         <json.version>20190722</json.version>
+        <carbon.analytics.version>3.0.36</carbon.analytics.version>
```
[the remark this answers is not in the document]
Comment (sahandilshan) [reply pom.xml]: Fixed in 3aac1ad2537b9d3e6462b70dbcde641c5e7513cb

File: component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/PollingMetrics.java (hunk 6 above)
Comment (sahandilshan) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/PollingMetrics.java]: Fixed in 9b553da8e0ca2f9d10e0868734562af8cfac30e7

File: component/pom.xml (hunk 7 above)
Comment (sahandilshan) [reply component/pom.xml]: Fixed in 9b553da8e0ca2f9d10e0868734562af8cfac30e7

File: component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/MetricsUtils.java [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+ * This class contains Util methods for the Metrics.
+ */
+public class MetricsUtils {
+
+    private static String getShortenedURL(String jdbcURL) {
+        if (jdbcURL.length() <= 30) {
+            return jdbcURL;
+        }
+        int n = jdbcURL.length();
+        int i = 30; // to get last 30 characters
```
Comment (senthuran16) [inline component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/MetricsUtils.java:30]: This should be `first 30 characters`, not `last` I guess?

Comment (sahandilshan) [reply component/src/main/java/io/siddhi/extension/io/cdc/source/metrics/MetricsUtils.java]: Resolved with 4286d561283b3cc311908e992f2db7d79ea4927e

**Pull request merged**